src/ILLink.RoslynAnalyzer/IMethodSymbolExtensions.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace ILLink.RoslynAnalyzer
+{
+	static class IMethodSymbolExtensions
+	{
+		public static bool IsSetter (this IMethodSymbol method)
+		{
+			return method.MethodKind == MethodKind.PropertySet;
+		}
+
+		public static bool IsGetter (this IMethodSymbol method)
+		{
+			return method.MethodKind == MethodKind.PropertyGet;
+		}
+		public static bool IsEventMethod (this IMethodSymbol method)
+		{
+			return method.MethodKind == MethodKind.EventAdd
+				|| method.MethodKind == MethodKind.EventRaise
+				|| method.MethodKind == MethodKind.EventRemove;
+		}
+		private static void PrependGenericParameters(ImmutableArray<ITypeParameterSymbol> genericParameters, System.Text.StringBuilder sb)
+		{
+			sb.Insert (0, '>').Insert (0, genericParameters[genericParameters.Length - 1]);
+			for (int i = genericParameters.Length - 2; i >= 0; i--)
+				sb.Insert (0, ',').Insert (0, genericParameters[i]);
+
+			sb.Insert (0, '<');
+		}
+
+		public static string GetDisplayName (this IMethodSymbol method)
```

Comment:
The GetDisplayName method was made to match C# naming which is what Roslyn uses and most people are used to reading instead of IL namings. Therefore if there is a discrepancy between the way it's represented by the analyzer and linker is likely a fix in the GetDisplayName in the linker not an implementation on the analyzer

Reply:
Worst case I think you can append directly the string ".cctor" at the end when displaying the diagnostic

Reply:
I would expect most cases to already be covered by Roslyn. I could understand that we need to do something special for property getter/setter (and events), and possibly even .cctor, but other than that I would expect to simply call Roslyn to do this for us.

Reply:
For now, would it be best to use the ISymbolExtensions.GetDisplayName and have separate ExpectedWarning's for the analyzer and trimmer, and then later change the trimmer's GetDisplayName to match the Analyzer/Rozlyn?

Reply:
I added a check in the GetDisplayName in the linker that exists for ctors, but was missing cctors. Now they should look the same.

Reply:
I think I was wrong is not a fix on the linker side, because if you fix in the linker then constructor and static constructor are represented equally, there is no way to differentiate them. So in the cases like the BeforeInitField test, I think things can get confusing. Same for RequiresInCompilerGeneratedCode tests, they get more confusing although you don't see the changes because we don't check for the caller name only messages in the attribute.
But yeah I think the best place to fix it is ISymbolExtensions.GetDisplayName an if statement before printing the method name check if the method symbol is a static constructor

Reply:
.cctors are weird - so we might need to add a special case unfortunately.

Reply:
I recently pushed a couple commits with different formats for the diagnostic messages. [The first one](https://github.com/dotnet/linker/pull/2455/commits/da05d3f6d36b153e20196a264daeb35c53c85fd4) includes the `static` modifier (e.g. `static Mono.Linker.Tests.Cases.RequiresCapability.RequiresCapability.StaticCtor.StaticCtor()`). [The second](https://github.com/dotnet/linker/pull/2455/commits/04270d67e306947a3d9a2aa82d65bf0f6325840b) uses the `..cctor()` notation (e.g. `Mono.Linker.Tests.Cases.RequiresCapability.RequiresCapability.StaticCtor..cctor()`).

I think the static modifier notation would be best since it still differentiates the constructors and is more familiar to most people.

Reply:
I just created something in SharpLab to see how a warning would be printed
```
public class Foo
{
    static Foo (int a) {}
}
```
This generates `error CS0132: 'Foo.Foo(int)': a static constructor must be parameterless` also inserting the [Obsolete] attribute will generate a message `[deprecated] Foo.Foo()` so the compiler warnings at this moment don't show the modifiers nor have a clear distinction between static and instance constructors, so whatever decision we make is going to deviate from whatever the compiler normally prints.

I think is key for the trimmer to have some differentiation but I don't have a preference for which one to choose
- The first option I have the concern that not a lot of people will even see the static modifier, it's at the beginning of the string and no other member will print it. Although I agree, I like more the readability of it.
- The second option is a weird notation, but static constructors are weird, we don't even allow them to annotate them. Static constructors are handled in a special way. So the fact that the name is weird might even be beneficial.